src/zenml/client.py

```diff
@@ -2386,6 +2392,113 @@ def _delete_artifact_metadata(
+                metadata was produced.
+            step_run_id: The ID of the step run during which the metadata was
+                produced.
+            artifact_id: The ID of the artifact for which the metadata was
+                produced.
+            stack_component_id: The ID of the stack component that produced
+                the metadata.
+
+        Returns:
+            The created metadata.
+
+        Raises:
+            ValueError: If not exactly one of either `pipeline_run_id`,
+                `step_run_id`, or `artifact_id` is provided.
+        """
+        from zenml.metadata.metadata_types import get_metadata_type
+
+        if not (pipeline_run_id or step_run_id or artifact_id):
+            raise ValueError(
+                "Cannot create run metadata without linking it to any entity. "
+                "Please provide either a `pipeline_run_id`, `step_run_id`, or "
+                "`artifact_id`."
+            )
+        if (
+            (pipeline_run_id and step_run_id)
+            or (pipeline_run_id and artifact_id)
+            or (step_run_id and artifact_id)
+        ):
+            raise ValueError(
+                "Cannot create run metadata linked to multiple entities. "
+                "Please provide only a `pipeline_run_id` or only a "
+                "`step_run_id` or only an `artifact_id`."
+            )
+
+        created_metadata_list: List[RunMetadataResponseModel] = []
+        for key, value in metadata.items():
+            if len(json.dumps(value)) > TEXT_FIELD_MAX_LENGTH:
+                logger.warning(
+                    f"Metadata value for key '{key}' is too large to be "
+                    "stored in the database. Skipping."
+                )
+                continue
+            run_metadata = RunMetadataRequestModel(
+                project=self.active_project.id,
+                user=self.active_user.id,
+                pipeline_run_id=pipeline_run_id,
+                step_run_id=step_run_id,
+                artifact_id=artifact_id,
+                stack_component_id=stack_component_id,
+                key=key,
+                value=value,
+                type=get_metadata_type(value),
+            )
+            created_metadata = self.zen_store.create_run_metadata(run_metadata)
+            created_metadata_list.append(created_metadata)
+        return created_metadata_list
+
+    def list_run_metadata(
+        self,
+        project_id: Optional[UUID] = None,
+        user_id: Optional[UUID] = None,
+        pipeline_run_id: Optional[UUID] = None,
+        step_run_id: Optional[UUID] = None,
+        artifact_id: Optional[UUID] = None,
+        stack_component_id: Optional[UUID] = None,
+    ) -> List[RunMetadataResponseModel]:
+        """List run metadata.
+
+        Args:
+            project_id: If provided, only return metadata for this project.
+            user_id: If provided, only return metadata for this user.
```

Comment:
Why is this scoped by user or project. a pipeline_run already is scoped by this.

Reply:
I imagine it could come handy if you want to see all metadata that a given user created, or all the metadata in a certain project. Now, how we resolve this in the backend is another question, I guess we could do it by joining the metadata schema with the respective entity that it is linked to (pipeline run / step run / artifact) and get the user/project from there, but I figured just adding it as foreign key directly makes that a lot easier

Reply:
alright makes sense, i mean it will move into the FIlter class from my PR anyways